File: packages/vite-plugin/src/share.ts
/**
 * Who is allowed to do what, when a deck is shared beyond this machine.
 *
 * The dev server writes to the author's files. Putting it on a network is
 * therefore a security decision and not a convenience, and this module is the
 * whole of that decision: nothing else in the plugin decides whether a request
 * may edit.
 *
 * # Four rules, and none of them is a default anyone can drift past
 *
 * **Nothing is shared unless somebody asked.** With no share secret in the
 * environment the routes behave exactly as they always have, on loopback, and
 * this module answers `Grant.Write` to every request — because the only thing
 * that can reach loopback is the author.
 *
 * **A link is a capability, and the secret travels in the fragment.** A
 * fragment is not sent with the request, so it reaches no access log, no
 * referrer header and no proxy record. The shape is
 * [`readPairing`](@slidxjs/runtime)'s, and the reader *is* `readPairing` — one
 * answer in this repository rather than two, and a URL that arrived with its
 * secret in the query is refused rather than honoured.
 *
 * **Read-only is the default; editing is a second secret.** Not a flag on the
 * same token — a different sixteen bytes, which a viewer was never given.
 * Somebody sharing their screen at a conference has not handed out the right to
 * rewrite their talk, and that is true by construction rather than by policy.
 *
 * **Loopback is always the author.** A request from `127.0.0.1` is this machine,
 * so the bookmark the author already had keeps working when sharing is switched
 * on. Everything else has to present a secret.
 */

import { timingSafeEqual } from "node:crypto";
import type { IncomingMessage, ServerResponse } from "node:http";

import { readPairing, type Pairing } from "@slidxjs/runtime";

/** What a request is allowed to do. */
export enum Grant {
  /** Not allowed to see the deck at all. */
  None = "none",
  /** May read the deck and say where in it they are. */
  Read = "read",
  /** May also change it. */
  Write = "write",
}

/**
 * Where the secrets come from.
 *
 * `slidx dev --crdt` mints them and passes them in the environment, so they
 * never touch the project directory and are not in the argument list of a
 * process anybody can list. The plugin can be used without slidx, in which case
 * these are unset and nothing is shared.
 */
export const SHARE_VARIABLE = "SLIDX_SHARE";
export const SHARE_EDIT_VARIABLE = "SLIDX_SHARE_EDIT";
/** Public LAN origin the CLI printed, used only to rebuild its links locally. */
export const SHARE_ORIGIN_VARIABLE = "SLIDX_SHARE_ORIGIN";

/** The header the editor presents its share credential in. */
export const CREDENTIAL_HEADER = "x-slidx-share";

/**
 * Browser-held reading access for the deck pages an authenticated editor embeds.
 *
 * It is a session cookie: closing the browser drops it, and stopping the dev
 * server invalidates the capability it contains. A cookie is always capped at
 * reading by [`grantForRequest`]; writes continue to require the explicit
 * header JavaScript derives from the URL fragment.
 */
export const CREDENTIAL_COOKIE = "slidx_share";

export interface Sharing {
  /** True when a secret was issued, which is the only time anything is shared. */
  readonly on: boolean;
  /** Links the local author may copy again. Never returned to a remote peer. */
  readonly links?: SharingLinks;
  /** What a request presenting this credential from this address may do. */
  grant(credential: string | undefined, address: string | undefined): Grant;
}

/** The capabilities the CLI printed when it started this shared session. */
export interface SharingLinks {
  readonly read?: string;
  readonly edit?: string;
}

/**
 * Reads the environment once.
 *
 * A malformed secret is treated as no secret at all rather than as a secret
 * nobody can match: half-configured sharing that silently refused every
 * request would look exactly like a network problem at a conference.
 */
export function createSharing(environment: NodeJS.ProcessEnv = process.env): Sharing {
  const read = pairing(environment[SHARE_VARIABLE]);
  const write = pairing(environment[SHARE_EDIT_VARIABLE]);
  const on = read !== null || write !== null;
  const links = sharingLinks(environment, read !== null, write !== null);

  return {
    on,
    ...(links ? { links } : {}),

    grant(credential, address) {
      if (!on || isLoopback(address)) return Grant.Write;

      const presented = pairing(credential);
      if (presented === null) return Grant.None;

      if (matches(presented, write)) return Grant.Write;
      if (matches(presented, read)) return Grant.Read;

      return Grant.None;
    },
  };
}

/**
 * Rebuilds the fragment links only when the CLI supplied their public origin.
 *
 * The server cannot infer this from a local editor request: its `Host` is
 * usually `localhost`, while the useful link names the machine on the LAN.
 * Rejecting paths, credentials and non-HTTP schemes keeps an environment typo
 * from turning the share sheet into a link to somewhere else.
 */
function sharingLinks(
  environment: NodeJS.ProcessEnv,
  hasRead: boolean,
  hasEdit: boolean,
): SharingLinks | undefined {
  const origin = shareOrigin(environment[SHARE_ORIGIN_VARIABLE]);
  if (origin === undefined) return undefined;

  const read = hasRead ? environment[SHARE_VARIABLE] : undefined;
  const edit = hasEdit ? environment[SHARE_EDIT_VARIABLE] : undefined;

  return {
    ...(read ? { read: `${origin}/__slidx/#s=${read}` } : {}),
    ...(edit ? { edit: `${origin}/__slidx/#s=${edit}` } : {}),
  };
}

function shareOrigin(value: string | undefined): string | undefined {
  if (!value) return undefined;

  try {
    const parsed = new URL(value);
    if (
      (parsed.protocol !== "http:" && parsed.protocol !== "https:") ||
      parsed.username ||
      parsed.password ||
      (parsed.pathname !== "/" && parsed.pathname !== "") ||
      parsed.search ||
      parsed.hash
    )
      return undefined;

    return parsed.origin;
  } catch {
    return undefined;
  }
}

/**
 * Access presented by one HTTP request.
 *
 * An explicit header retains its full grant. A browser cookie is deliberately
 * capped at reading so another page on the same site cannot turn an ambient
 * credential into a file-writing request.
 */
export function grantForRequest(sharing: Sharing, request: IncomingMessage): Grant {
  const header = headerCredential(request);
  if (header !== undefined) return sharing.grant(header, request.socket.remoteAddress);

  const cookie = cookieCredential(request);
  const grant = sharing.grant(cookie, request.socket.remoteAddress);
  return cookie !== undefined && grant === Grant.Write ? Grant.Read : grant;
}

/** The explicit capability the editor presents after reading its fragment. */
export function headerCredential(request: IncomingMessage): string | undefined {
  const presented = request.headers[CREDENTIAL_HEADER];
  return Array.isArray(presented) ? presented[0] : presented;
}

/**
 * Lets subsequent same-origin slide and asset requests carry reading access.
 *
 * Only called after the header was accepted. `HttpOnly` keeps the credential
 * out of deck scripts, and `SameSite=Strict` keeps it off cross-site requests.
 */
export function rememberReadAccess(request: IncomingMessage, response: ServerResponse): void {
  const presented = headerCredential(request);
  if (presented === undefined) return;

  response.setHeader(
    "set-cookie",
    `${CREDENTIAL_COOKIE}=${encodeURIComponent(presented)}; Path=/; HttpOnly; SameSite=Strict`,
  );
}

function cookieCredential(request: IncomingMessage): string | undefined {
  const cookies = request.headers.cookie;
  if (!cookies) return undefined;

  const value = cookies
    .split(";")
    .map((part) => part.trim())
    .find((part) => part.startsWith(`${CREDENTIAL_COOKIE}=`))
    ?.slice(CREDENTIAL_COOKIE.length + 1);
  if (value === undefined) return undefined;

  try {
    return decodeURIComponent(value);
  } catch {
    return undefined;
  }
}

/**
 * One credential, read the way a phone remote's pairing is read.
 *
 * The value is the fragment's own text, so it is handed back to `readPairing`
 * with the `#` it was found after. That keeps the refusal of a secret in a
 * query string in one place — this module never learns the rule.
 */
function pairing(value: string | undefined): Pairing | null {
  if (value === undefined || value.length === 0) return null;

  return readPairing(`#s=${value}`);
}

/**
 * Whether two pairings are the same, without leaking how far they matched.
 *
 * A capability guarded by nothing but its own length has to be compared in
 * constant time: a byte-at-a-time comparison tells an attacker on the same
 * network which prefix was right, and sixteen bytes guessed one byte at a time
 * is not sixteen bytes.
 */
function matches(presented: Pairing, issued: Pairing | null): boolean {
  if (issued === null) return false;

  return same(presented.session, issued.session) && same(presented.secret, issued.secret);
}

function same(a: string, b: string): boolean {
  const left = Buffer.from(a, "utf8");
  const right = Buffer.from(b, "utf8");

  // `timingSafeEqual` throws on a length mismatch, which is itself a leak of
  // one bit. A length that differs cannot be the issued secret anyway, because
  // slidx issues one length.
  return left.length === right.length && timingSafeEqual(left, right);
}

/**
 * True for the machine the dev server is running on.
 *
 * Node reports an IPv4 peer over a dual-stack socket as `::ffff:127.0.0.1`, so
 * the mapped form has to be recognised too — missing it would lock the author
 * out of their own editor the moment they shared it.
 */
export function isLoopback(address: string | undefined): boolean {
  if (address === undefined) return false;

  const plain = address.replace(/^::ffff:/, "");

  return plain === "127.0.0.1" || plain === "::1" || plain.startsWith("127.");
}

File: crates/slidx_cli/src/dev/share.rs
//! The share link, and every decision that makes it safe to print.
//!
//! `slidx dev --crdt` puts a dev server that can write the author's files onto a
//! network. That is a security decision, so it is one module rather than a flag
//! read in four places, and the four rules it keeps are the ones a convenient
//! default gets wrong.
//!
//! **The local network, not the internet.** A URL a co-presenter opens on the
//! same Wi-Fi is what a conference needs, and it involves no third party at all.
//! slidx offers **no tunnel**, and there is no flag that adds one. A public URL
//! to an unannounced talk, served by a process that can write the author's files,
//! through a third party that then sees the traffic, is a different feature with
//! a different blast radius — and the version of it that is one flag away from
//! `--crdt` is the version somebody reaches for by accident. `slidx dev --help`
//! says so, because a decision nobody can read is one somebody will file as a
//! bug.
//!
//! **Bound beyond localhost only when sharing was asked for.** Without `--crdt`,
//! Vite is started with no `--host` and the deck is reachable from this machine
//! and nowhere else, exactly as before.
//!
//! **The secret is in the fragment.** After the `#`, never in the query. A
//! fragment is not sent with the request, so it reaches no access log, no
//! referrer header and no proxy record. The shape is `pairingUrl`'s from
//! `packages/runtime/src/remote.ts`, and the *reader* on the other side is that
//! file's own `readPairing` — so a URL slidx prints and a URL the editor accepts
//! cannot drift apart without a test failing.
//!
//! **Read-only unless editing was granted separately.** Two secrets, not one
//! secret and a flag. A viewer holding the read link cannot reach the edit route
//! because they were never given the bytes that open it, which is a property
//! rather than a policy.

pub mod address;
pub mod qr;
pub mod secret;

use std::net::IpAddr;

use crate::report;
use crate::style::{Ink, Style};

use secret::{NoRandomness, SECRET_BYTES, SESSION_BYTES};

/// The fragment key the secret travels under.
///
/// `FRAGMENT_KEY` in `packages/runtime/src/remote.ts`, so other fragment
/// parameters can coexist with it. Changing it here alone would print links the
/// editor reads nothing out of.
const FRAGMENT_KEY: &str = "s";

/// The editor's route, which is what a share link points at.
const EDITOR_ROUTE: &str = super::EDITOR_ROUTE;

/// The environment variables the dev server reads the secrets out of.
///
/// The environment rather than the command line, because an argument list is
/// readable by every process on the machine and an environment is not. Named the
/// same as `SHARE_VARIABLE` and `SHARE_EDIT_VARIABLE` in
/// `packages/vite-plugin/src/share.ts`, which is the only reader.
pub const SHARE_VARIABLE: &str = "SLIDX_SHARE";
pub const SHARE_EDIT_VARIABLE: &str = "SLIDX_SHARE_EDIT";
/// Public LAN origin the plugin uses to rebuild the printed links for the author.
pub const SHARE_ORIGIN_VARIABLE: &str = "SLIDX_SHARE_ORIGIN";

/// One shared session's capabilities.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Share {
    /// Names the session. Not secret, and it appears in every link.
    pub session: String,
    /// Opens the deck for reading.
    pub read: String,
    /// Opens it for editing. `None` unless `--allow-edit` was given.
    pub edit: Option<String>,
}

impl Share {
    /// Mints a session, and an edit secret only if one was asked for.
    ///
    /// The edit secret is not minted and withheld — it does not exist. A secret
    /// that was generated and then not printed is still a secret in a process's
    /// environment, and the honest shape of "nobody may edit" is that there is
    /// nothing to present.
    pub fn mint(allow_edit: bool) -> Result<Self, NoRandomness> {
        Ok(Self {
            session: secret::token(SESSION_BYTES)?,
            read: secret::token(SECRET_BYTES)?,
            edit: if allow_edit { Some(secret::token(SECRET_BYTES)?) } else { None },
        })
    }

    /// The credential for one secret, as the environment and the fragment spell it.
    pub fn credential(&self, secret: &str) -> String {
        format!("{}.{secret}", self.session)
    }

    /// The link to hand somebody, with the secret after the `#`.
    pub fn link(&self, origin: &str, secret: &str) -> String {
        format!("{origin}{EDITOR_ROUTE}#{FRAGMENT_KEY}={}", self.credential(secret))
    }

    /// The variables to start the dev server with.
    pub fn environment(&self) -> Vec<(&'static str, String)> {
        let mut variables = vec![(SHARE_VARIABLE, self.credential(&self.read))];

        if let Some(edit) = &self.edit {
            variables.push((SHARE_EDIT_VARIABLE, self.credential(edit)));
        }

        variables
    }
}

/// What is printed under the ready line when a deck is shared.
///
/// The read link, its code, and — only when editing was granted — the second
/// link and its own code. Then the two sentences somebody needs before they read
/// the URL out to a room.
pub fn block(share: &Share, address: IpAddr, port: u16, indent: usize, style: &Style) -> String {
    let origin = address::origin(address, port);
    let mut text = String::new();

    text.push('\n');
    text.push_str(&report::flowed("read only — anyone with this link", indent, Ink::Faint, style));
    text.push_str(&link(&share.link(&origin, &share.read), indent, style));

    if let Some(edit) = &share.edit {
        text.push('\n');
        text.push_str(&report::flowed(
            "can edit — this one rewrites your slide files",
            indent,
            Ink::Warn,
            style,
        ));
        text.push_str(&link(&share.link(&origin, edit), indent, style));
    }

    text.push('\n');
    text.push_str(&report::flowed(caution(share), indent, Ink::Faint, style));

    text
}

/// One link, and its code underneath.
fn link(url: &str, indent: usize, style: &Style) -> String {
    let mut text = report::flowed(url, indent, Ink::Strong, style);
    let code = qr::block(url, indent, style);

    if !code.is_empty() {
        text.push('\n');
        text.push_str(&code);
    }

    text
}

/// The sentence that has to be read before the URL is.
fn caution(share: &Share) -> &'static str {
    match share.edit {
        Some(_) => {
            "Both links work for anyone on this network who has them, and the second one \
                    can change your deck. They stop working when this dev server does."
        }
        None => {
            "This link works for anyone on this network who has it. It cannot change your \
                 deck, and it stops working when this dev server does."
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::net::Ipv4Addr;

    fn share() -> Share {
        Share {
            session: "0123456789abcdef".into(),
            read: "00112233445566778899aabbccddeeff".into(),
            edit: Some("ffeeddccbbaa99887766554433221100".into()),
        }
    }

    fn here() -> IpAddr {
        IpAddr::V4(Ipv4Addr::new(192, 168, 1, 42))
    }

    #[test]
    fn a_link_carries_the_secret_after_the_hash_and_nowhere_else() {
        // The single character this whole module exists for. A query parameter
        // reaches an access log, and a log holding this value is enough to join.
        let url = share().link("http://192.168.1.42:5173", &share().read);

        assert_eq!(
            url,
            "http://192.168.1.42:5173/__slidx/#s=0123456789abcdef.00112233445566778899aabbccddeeff"
        );
        assert!(!url.contains('?'), "{url}");
        assert!(url.split('#').next().is_some_and(|before| !before.contains(&share().read)));
    }

    #[test]
    fn a_links_fragment_is_shaped_the_way_the_phone_remote_reads_one() {
        // `readPairing` in packages/runtime/src/remote.ts is the reader, and it
        // wants `s=<session>.<secret>` in lowercase hex. This is the pin on the
        // Rust side; `test/share.test.ts` is the pin on the other.
        let credential = share().credential(&share().read);

        assert_eq!(credential, "0123456789abcdef.00112233445566778899aabbccddeeff");
        assert!(credential.split('.').all(|part| part.chars().all(|c| c.is_ascii_hexdigit())));
    }

    #[test]
    fn a_session_without_edit_access_has_no_edit_secret_at_all() {
        // Not minted and withheld — absent. A secret in a process's environment
        // is a secret whether or not it was printed.
        let read_only = Share::mint(false).expect("randomness");

        assert!(read_only.edit.is_none());
        assert_eq!(read_only.environment().len(), 1);
        assert_eq!(read_only.environment()[0].0, SHARE_VARIABLE);
    }

    #[test]
    fn granting_edit_access_adds_a_second_secret_rather_than_changing_the_first() {
        let both = Share::mint(true).expect("randomness");

        assert_ne!(both.edit.as_deref(), Some(both.read.as_str()));
        assert_eq!(both.environment().len(), 2);
    }

    #[test]
    fn the_variables_are_the_ones_the_plugin_reads() {
        // Named in packages/vite-plugin/src/share.ts, which is the only reader.
        assert_eq!(SHARE_VARIABLE, "SLIDX_SHARE");
        assert_eq!(SHARE_EDIT_VARIABLE, "SLIDX_SHARE_EDIT");
    }

    #[test]
    fn a_read_only_share_says_the_link_cannot_change_the_deck() {
        let text = block(&Share { edit: None, ..share() }, here(), 5173, 2, &Style::plain());

        assert!(text.contains("read only"), "{text}");
        assert!(text.contains("cannot change your"), "{text}");
        assert!(!text.contains("can edit"), "{text}");
    }

    #[test]
    fn a_share_that_grants_editing_warns_about_the_link_that_does_it() {
        // Somebody about to read a URL out to a room needs to know which of the
        // two they are reading.
        let text = block(&share(), here(), 5173, 2, &Style::plain());

        assert!(text.contains("can edit"), "{text}");
        assert!(text.contains("rewrites your slide files"), "{text}");
        assert!(text.contains(&share().read), "{text}");
        assert!(text.contains(share().edit.as_deref().unwrap()), "{text}");
    }

    #[test]
    fn a_share_block_says_the_links_end_with_the_dev_server() {
        // The one reassurance that is actually true: nothing outlives the
        // process, because nothing else is holding the secret.
        assert!(
            block(&share(), here(), 5173, 2, &Style::plain()).contains("when this dev server does")
        );
    }

    #[test]
    fn a_share_block_carries_no_escape_sequences_when_colour_is_off() {
        assert!(!block(&share(), here(), 5173, 2, &Style::plain()).contains('\u{1b}'));
    }

    #[test]
    fn a_coloured_share_block_draws_the_code_under_the_link() {
        let text = block(&share(), here(), 5173, 2, &Style::colored());

        assert!(text.contains('\u{2588}') || text.contains('\u{2580}'), "no code drawn");
    }
}
